cedar-policy-core/src/parser/cst_to_ast.rs

```diff
@@ -1096,403 +1027,325 @@ impl RefKind for OneOrMultipleRefs {
         "an entity uid or set of entity uids"
     }
 
-    fn create_slot(errs: &mut ToASTErrors, loc: &Loc) -> Option<Self> {
-        errs.push(ToASTError::new(
+    fn create_slot(loc: &Loc) -> Result<Self> {
+        Err(ToASTError::new(
             RefCreationError::two_expected(Ref::Single, Ref::Set, Ref::Template).into(),
             loc.clone(),
-        ));
-        None
+        )
+        .into())
     }
 
-    fn create_single_ref(e: EntityUID, _errs: &mut ToASTErrors, _loc: &Loc) -> Option<Self> {
-        Some(OneOrMultipleRefs::Single(e))
+    fn create_single_ref(e: EntityUID, _loc: &Loc) -> Result<Self> {
+        Ok(OneOrMultipleRefs::Single(e))
     }
 
-    fn create_multiple_refs(
-        es: Vec<EntityUID>,
-        _errs: &mut ToASTErrors,
-        _loc: &Loc,
-    ) -> Option<Self> {
-        Some(OneOrMultipleRefs::Multiple(es))
+    fn create_multiple_refs(es: Vec<EntityUID>, _loc: &Loc) -> Result<Self> {
+        Ok(OneOrMultipleRefs::Multiple(es))
     }
 }
 
 impl Node<Option<cst::Or>> {
-    fn to_expr_or_special(&self, errs: &mut ToASTErrors) -> Option<ExprOrSpecial<'_>> {
-        // if `self` doesn't have data, nothing we can do here, just propagate
-        // the `None`; we don't need to signal an error, because one was already
-        // signaled when the `Node` without data was created
-        let or = self.as_inner()?;
-
-        let maybe_first = or.initial.to_expr_or_special(errs);
-        let mut more = or.extended.iter().filter_map(|i| i.to_expr(errs));
-        // getting the second here avoids the possibility of a singleton construction
-        let maybe_second = more.next();
-        // collect() preforms all the conversions, generating any errors
-        let rest: Vec<_> = more.collect();
-
-        match (maybe_first, maybe_second, rest.len(), or.extended.len()) {
-            (f, None, _, 0) => f,
-            (Some(f), Some(s), r, e) if 1 + r == e => {
```

Comment:
The condition was checking if `.filter_map(|i| i.to_expr(errs))` ended up filtering anything out of the list. In case it filtered anything out (i.e., `1 + r < e`), the previous code would return `None` (because presumably something had been added to `errs`). In the rewrite, this logic is handled by `ParseErrors::transpose` and `flatten_tuple_2`.